internal/state/relationship/capturer.go

```diff
@@ -63,32 +84,89 @@ func (c *CapturerImpl) Capture(obj client.Object) {
 				Name:      svcName,
 			}
 		}
+	case *v1beta1.Gateway:
+		var selectors []labels.Selector
+		for _, listener := range o.Spec.Listeners {
+			if selector := graph.GetAllowedRouteLabelSelector(listener); selector != nil {
+				convertedSelector, err := metav1.LabelSelectorAsSelector(selector)
+				if err == nil {
+					selectors = append(selectors, convertedSelector)
+				}
+			}
+		}
+
+		nsname := client.ObjectKeyFromObject(o)
+		if len(selectors) > 0 {
+			c.gatewayLabelSelectors[nsname] = selectors
+			for ns, cfg := range c.referencedNamespaces {
+				for _, selector := range selectors {
+					if selector.Matches(labels.Set(cfg.labelMap)) {
+						cfg.match = true
+						cfg.gateways[nsname] = struct{}{}
+						break
+					}
+				}
+				c.referencedNamespaces[ns] = cfg
+			}
+		} else if _, exists := c.gatewayLabelSelectors[nsname]; exists {
+			// label selectors existed previously for this gateway, so clean up any references to them
+			c.removeGatewayLabelSelector(nsname)
+		}
+	case *v1.Namespace:
+		nsLabels := o.GetLabels()
+		gateways := c.matchingGateways(nsLabels)
+		nsCfg := namespaceCfg{
+			labelMap: nsLabels,
+			match:    len(gateways) > 0,
+			gateways: gateways,
+		}
+		c.referencedNamespaces[client.ObjectKeyFromObject(o)] = nsCfg
 	}
 }
 
 // Remove removes the relationship for the given object from the CapturerImpl.
-func (c *CapturerImpl) Remove(resourceType client.Object, nsname types.NamespacedName) {
-	switch resourceType.(type) {
+func (c *CapturerImpl) Remove(resource client.Object, nsname types.NamespacedName) {
+	switch resource.(type) {
 	case *v1beta1.HTTPRoute:
 		c.deleteForRoute(nsname)
 	case *discoveryV1.EndpointSlice:
 		delete(c.endpointSliceOwners, nsname)
+	case *v1beta1.Gateway:
+		c.removeGatewayLabelSelector(nsname)
+	case *v1.Namespace:
+		delete(c.referencedNamespaces, nsname)
 	}
 }
 
 // Exists returns true if the given object has a relationship with another object.
-func (c *CapturerImpl) Exists(resourceType client.Object, nsname types.NamespacedName) bool {
-	switch resourceType.(type) {
+func (c *CapturerImpl) Exists(resource client.Object, nsname types.NamespacedName) bool {
+	switch resource.(type) {
 	case *v1.Service:
 		return c.serviceRefCount[nsname] > 0
 	case *discoveryV1.EndpointSlice:
 		svcOwner, exists := c.endpointSliceOwners[nsname]
 		return exists && c.serviceRefCount[svcOwner] > 0
+	case *v1.Namespace:
+		cfg, exists := c.referencedNamespaces[nsname]
+		return exists && cfg.match
 	}
 
 	return false
 }
 
+// RelationshipEnded checks if a relationship previously existed, but no longer does.
+func (c CapturerImpl) RelationshipEnded(resource client.Object) bool {
+	switch o := resource.(type) {
+	case *v1.Namespace:
```

Comment:
it looks like every subsequent invocation of this code will always return true, meaning it will always report relationship ended. Is it correct?

Reply:
When we call Capture (after calling this function), it would remove the relationship, and `c.Exists` would return false.